### PR TITLE
Point out shadowed associated types on E0191

### PR DESCRIPTION
**Pull request by TheLazyDutchman**
Previously, when you tried to set an associated type that is shadowed by an associated type in a subtrait, like this:
```rust
trait A {
    type X;
}

trait B: A {
    type X; // note: this is legal
}

impl<Y> Clone for Box<dyn B<X=Y, X=Y>> {
    fn clone(&self) -> Self {
        todo!()
    }
}
```
you got a confusing error message, that says nothing about the shadowing:
```rust
error[E0719]: the value of the associated type `X` (from trait `B`) is already specified
 --> test.rs:9:34
  |
9 | impl<Y> Clone for Box<dyn B<X=Y, X=Y>> {
  |                             ---  ^^^ re-bound here
  |                             |
  |                             `X` bound here first

error[E0191]: the value of the associated type `X` (from trait `A`) must be specified
 --> test.rs:9:27
  |
2 |     type X;
  |     ------ `X` defined here
...
9 | impl<Y> Clone for Box<dyn B<X=Y, X=Y>> {
  |                           ^^^^^^^^^^^ help: specify the associated type: `B<X=Y, X=Y, X = Type>`

error: aborting due to 2 previous errors

Some errors have detailed explanations: E0191, E0719.
For more information about an error, try `rustc --explain E0191`.
```
Now instead, the error shows that the associated type is shadowed, and suggests renaming as a potential fix.
```rust
error[E0719]: the value of the associated type `X` in trait `B` is already specified
 --> test.rs:9:34
  |
9 | impl<Y> Clone for Box<dyn B<X=Y, X=Y>> {
  |                             ---  ^^^ re-bound here
  |                             |
  |                             `X` bound here first

error[E0191]: the value of the associated type `X` in `A` must be specified
 --> test.rs:9:27
  |
2 |     type X;
  |     ------ `A::X` defined here
...
6 |     type X; // note: this is legal
  |     ------ `A::X` shadowed here
...
9 | impl<Y> Clone for Box<dyn B<X=Y, X=Y>> {
  |                           ^^^^^^^^^^^ associated type `X` must be specified
  |
help: consider renaming this associated type
 --> test.rs:2:5
  |
2 |     type X;
  |     ^^^^^^
help: consider renaming this associated type
 --> test.rs:6:5
  |
6 |     type X; // note: this is legal
  |     ^^^^^^

error: aborting due to 2 previous errors

Some errors have detailed explanations: E0191, E0719.
For more information about an error, try `rustc --explain E0191`.
```

The rename help message is only emitted when the trait is local. This is true both for the supertrait as for the subtrait.

There might be cases where you can use the fully qualified path (for instance, in a where clause), but this PR currently does not deal with that.

### CODE REVIEW

**Comment by TheLazyDutchman** [comment]
fixes #100109

**Comment by compiler-errors** [comment]
Thanks for the contribution! This needs a UI test. 

Also, can you amend the PR title to have some sort of verb? "[Point out] overlapping associated types [when...?]" or something is more conventional title for a PR like this.

Also, can you squash the commits into one?

@rustbot author

File: compiler/rustc_hir_analysis/src/astconv/errors.rs [hunk 1]
```diff
@@ -585,6 +586,24 @@ impl<'o, 'tcx> dyn AstConv<'tcx> + 'o {
             }
         }
 
+        let bound_names = trait_bounds
```
Comment (compiler-errors) [inline compiler/rustc_hir_analysis/src/astconv/errors.rs:589]: Can you explain what this code is doing? Some comments throuhout may be helpful.

File: compiler/rustc_hir_analysis/src/astconv/errors.rs [hunk 2]
```diff
@@ -585,6 +586,24 @@ impl<'o, 'tcx> dyn AstConv<'tcx> + 'o {
...
+            .iter()
+            .flat_map(|poly_trait_ref| {
+                poly_trait_ref.trait_ref.path.segments.iter().flat_map(|x| {
+                    x.args.iter().flat_map(|args| {
```
Comment (compiler-errors) [inline compiler/rustc_hir_analysis/src/astconv/errors.rs:593]: Why are you looking at all of the segments in the trait ref? That shouldn't be necessary. Only the final path segment should be interesting here.

Comment (compiler-errors) [reply compiler/rustc_hir_analysis/src/astconv/errors.rs:593]: Also, please don't `iter()` over an `Option` -- `if let` or `let .. else` should work here more effectively.

File: compiler/rustc_hir_analysis/src/astconv/errors.rs [hunk 3]
```diff
@@ -585,6 +586,24 @@ impl<'o, 'tcx> dyn AstConv<'tcx> + 'o {
...
+                        args.bindings.iter().map(|binding| {
+                            let name = binding.ident.name;
+                            let trait_def = poly_trait_ref.trait_ref.path.res.opt_def_id();
+                            let assoc_item = trait_def.and_then(|did| {
+                                tcx.associated_items(did).filter_by_name_unhygienic(name).next()
```
Comment (compiler-errors) [inline compiler/rustc_hir_analysis/src/astconv/errors.rs:598]: We shouldn't be filtering unhygenically. You have an identifier in `binding.ident`.

**Review by compiler-errors — CHANGES_REQUESTED** [review]

**Comment by TheLazyDutchman** [comment]
continues in #117661, because I didn't know renaming the branch would close the PR

**Pull request closed without merging**